api-server/api-server-common/src/storage/impls/postgres/transactional/mod.rs

```diff
@@ -30,171 +30,182 @@ use crate::storage::storage_api::{
-            conn.get_block(block_id)
-        })?;
+        let mut conn = QueryFromConnection::new(&self.connection);
+        let res = conn.get_block(block_id).await?;
 
         Ok(res)
     }
 
     #[allow(clippy::type_complexity)]
-    pub fn get_transaction(
+    pub async fn get_transaction(
         &mut self,
         transaction_id: Id<Transaction>,
     ) -> Result<Option<(Option<Id<Block>>, SignedTransaction)>, ApiServerStorageError> {
-        let res = self.with_transaction_mut(|tx| {
-            let mut conn = QueryFromConnection::new(tx);
-            conn.get_transaction(transaction_id)
-        })?;
+        let mut conn = QueryFromConnection::new(&self.connection);
+        let res = conn.get_transaction(transaction_id).await?;
 
         Ok(res)
     }
 
-    pub fn get_block_aux_data(
+    pub async fn get_block_aux_data(
         &mut self,
         block_id: Id<Block>,
     ) -> Result<Option<BlockAuxData>, ApiServerStorageError> {
-        let res = self.with_transaction_mut(|tx| {
-            let mut conn = QueryFromConnection::new(tx);
-            conn.get_block_aux_data(block_id)
-        })?;
+        let mut conn = QueryFromConnection::new(&self.connection);
+        let res = conn.get_block_aux_data(block_id).await?;
 
         Ok(res)
     }
 }
 
-#[ouroboros::self_referencing]
-pub struct ApiServerPostgresTransactionalRw {
-    connection: PooledConnection<PostgresConnectionManager<NoTls>>,
-    #[borrows(mut connection)]
-    // TODO(PR): Dear reviewer, is the lifetime below covariant? It does look covariant to me.
-    #[covariant]
-    // Note: This Option is required to make destruction by value possible. See commit() method.
-    transaction: Option<postgres::Transaction<'this>>,
+pub struct ApiServerPostgresTransactionalRw<'a> {
+    connection: PooledConnection<'a, PostgresConnectionManager<NoTls>>,
+    finished: bool,
 }
 
-impl ApiServerPostgresTransactionalRw {
-    pub(super) fn from_connection(
-        connection: PooledConnection<PostgresConnectionManager<NoTls>>,
-    ) -> Result<Self, ApiServerStorageError> {
-        let tx = ApiServerPostgresTransactionalRwTryBuilder {
-            connection,
-            transaction_builder: |conn| {
-                conn.build_transaction()
-                    .read_only(true)
-                    .start()
-                    .map(Some)
-                    .map_err(|e| ApiServerStorageError::RwTxBeginFailed(e.to_string()))
-            },
+impl<'a> Drop for ApiServerPostgresTransactionalRw<'a> {
+    fn drop(&mut self) {
+        if !self.finished {
+            futures::executor::block_on(self.connection.batch_execute("ROLLBACK")).unwrap_or_else(
+                |e| {
+                    logging::log::error!(
+                        "CRITICAL ERROR: failed to rollback failed postgres RW transaction: {e}"
+                    )
+                },
+            );
         }
-        .try_build()?;
+    }
```

Comment:
I am not sure about all the implications of `block_on` in the destructor, seems like it could lead to problems if the future needs something else to resolve as part of its execution. Also, if I understand this correctly, the executor from the `futures` crate is different from the tokio one. Is it fine to mix and match them?

Not sure what to do about it. One option is to enforce that the user uses explicitly one of the `commit` or `rollback` methods to end the transaction. In the destructor, we just assert `self.finished`. However, it may be hard to chase up all the places where a transaction object is destroyed, especially considering all the `?`s and other forms of early return.

Reply:
Is there a test that exercises this code path?

Reply:
There's zero tests for postgres yet. Alfie is working on the containerization.

I've been thinking about this block_on problem yesterday all night (and panicking when forgetting occurred to me, but it's ugly)... I have an idea on how to get rid of it here... I'll discuss it in the daily.

Reply:
All my attempts to move the processing to the db as a parent failed so far. There's always the requirement to make something here or there 'static. The closest I came to succeeding is to make channels that send tasks to a tokio task, but then because the connection has a lifetime, it's mandating having 'static lifetime for the receiver.

scoped tokio didn't help because it returns a reference that cannot be moved out of scope.

Same for std scoped threads, because the scope cannot be moved.

No idea how I'll proceed. Maybe I'll have the patience to do it later. Feel free to take a poke at it.